src/systems/systems_wasm/src/aspects.rs

```diff
@@ -0,0 +1,852 @@
+    }
+
+    /// Write a set of variables to an object.
+    pub fn variables_to_object(&mut self, instance: Box<dyn VariablesToObject>) {
+        let variables = self.variables.register_many(&instance.variables());
+
+        self.actions.push((
+            ToObject::new(instance, variables).into(),
+            ExecuteOn::PostTick,
+        ));
+    }
+
+    /// Convert event occurrences to a variable.
+    ///
+    /// If you want to write a variable back to the same event, then use the
+    /// event id returned by this method and pass it to [Self::variable_to_event_id].
+    pub fn event_to_variable(
+        &mut self,
+        event_name: &str,
+        mapping: EventToVariableMapping,
+        target: Variable,
+        configure_options: fn(EventToVariableOptions) -> EventToVariableOptions,
+    ) -> Result<u32, Box<dyn Error>> {
+        Self::precondition_not_aircraft_variable(&target);
+
+        let target = self.variables.register(&target);
+
+        let event_to_variable = EventToVariable::new(
+            &mut self.sim_connect,
+            event_name,
+            mapping,
+            target,
+            configure_options(EventToVariableOptions::default()),
+        )?;
+
+        let event_id = event_to_variable.event_id;
+
+        self.message_handlers.push(event_to_variable.into());
+
+        Ok(event_id)
+    }
+
+    /// Write the variable's value to an event. If you use [Self::event_to_variable] for the same
+    /// event, then you should use [Self::variable_to_event_id] instead.
+    pub fn variable_to_event(
+        &mut self,
+        input: Variable,
+        mapping: VariableToEventMapping,
+        write_on: VariableToEventWriteOn,
+        event_name: &str,
+    ) -> Result<(), Box<dyn Error>> {
+        let input = self.variables.register(&input);
+
+        self.actions.push((
+            ToEvent::new(&mut self.sim_connect, input, mapping, write_on, event_name)?.into(),
+            ExecuteOn::PostTick,
+        ));
+
+        Ok(())
+    }
+
+    /// Write the variable's value to an event with the given event id. This function should be used
+    /// when you previously acquired an event id using [Self::event_to_variable].
+    pub fn variable_to_event_id(
+        &mut self,
+        input: Variable,
+        mapping: VariableToEventMapping,
+        write_on: VariableToEventWriteOn,
+        event_id: u32,
+    ) {
+        let input = self.variables.register(&input);
+
+        self.actions.push((
+            ToEvent::new_with_event_id(input, mapping, write_on, event_id).into(),
+            ExecuteOn::PostTick,
+        ));
+    }
+
+    fn precondition_not_aircraft_variable(variable: &Variable) {
+        if matches!(variable, Variable::Aircraft(..)) {
+            eprintln!("Writing to variable '{}' is unsupported.", variable);
+        }
+    }
+}
+
+pub struct MsfsAspect {
+    message_handlers: Vec<MessageHandler>,
+    actions: Vec<(VariableAction, ExecuteOn)>,
+}
```

Comment:
It is a bit unclear to me what the difference between a message handler and an action is. It seems like a message handler is an action to handle SimConnect events while an action is typically some generic variable/event mapping.

Is there a fundamental reason why we cannot treat message handlers as an action as well?

Reply:
Yes, the fundamental reason is that `actions` are executed every frame (either before or after a simulation tick), whereas `message_handlers` only handle things when the SimConnect message it handles is received.